### PR TITLE
DRAFT: WAM preview for console app

### PR DESCRIPTION
**Pull request by v-michaelmi**
## Purpose

This is meant to be a sample PR for using WAM instead of IWA. Will likely move to its own REPO soon.

### CODE REVIEW

File: iwa-console/MSAL/MSALClientHelper.cs [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+
+            await AttachTokenCache();
+            return await FetchSignedInUserFromCache().ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Attaches the token cache to the Public Client app.
+        /// </summary>
+        /// <returns>IAccount list of already signed-in users (if available)</returns>
+        private async Task<IEnumerable<IAccount>> AttachTokenCache()
```
Comment (bgavrilMS) [inline iwa-console/MSAL/MSALClientHelper.cs:108]: This method is doing too much, consider using just "AttachToCache"

File: iwa-console/appsettings.json [hunk 2]
```diff
@@ -1,28 +1,14 @@
...
-
-    // ClientId (ApplicationId) as copied from the application registration (which depends on the cloud instance)
-    // See docs referenced in the AzureCloudInstance section above
-    "ClientId": "ebe2ab4d-12b3-4446-8480-5c3828d04c50"
+{
+  "AzureAd": {
+    "Authority": "https://login.microsoftonline.com/{0}",
+    "TenantId": "[Enter Tenant Id (Obtained from the Azure portal. Select 'Endpoints' from the 'App registrations' blade and use the GUID in any of the URLs), e.g. da41245a5-11b3-996c-00a8-4d99re19f292]",
+    "ClientId": "[Enter the Client Id of the service (Application ID obtained from the Azure portal), e.g. ba74781c2-53c2-442a-97c2-3d60re42f403]",
+    "RedirectURI": "ms-appx-web://microsoft.aad.brokerplugin/{0}",
```
Comment (bgavrilMS) [inline iwa-console/appsettings.json:6]: The redirect uri should be "http://localhost" because this is what MSAL.NET uses on Mac and Linux and old Windows. 
`"ms-appx-web://microsoft.aad.brokerplugin/{0}"` only needs to be set at app registration level.

File: iwa-console/iwa-console.csproj [hunk 3]
```diff
@@ -2,24 +2,32 @@
...
-    <None Update="appsettings.json">
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
+    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" Version="7.0.0" />
+    <PackageReference Include="Microsoft.Extensions.Configuration.Json" Version="7.0.0" />
+    <PackageReference Include="Microsoft.Graph" Version="4.48.0" />
+    <PackageReference Include="Microsoft.Identity.Client" Version="4.48.1" />
+    <PackageReference Include="Microsoft.Identity.Client.Broker" Version="4.48.1-preview" />
+    <PackageReference Include="Microsoft.Identity.Client.Extensions.Msal" Version="2.25.0" />
+    <PackageReference Include="Microsoft.Windows.CppWinRT" Version="2.0.221121.5" />
```
Comment (bgavrilMS) [inline iwa-console/iwa-console.csproj:25]: Why is this needed?

File: iwa-console/iwa-console.csproj [hunk 4]
```diff
@@ -2,24 +2,32 @@
...
+    <PackageReference Include="Newtonsoft.Json" Version="13.0.2" />
```
Comment (bgavrilMS) [inline iwa-console/iwa-console.csproj:26]: Do not use newtonsoft anymore. Use System.Text.Json as per .NET guidelines - https://learn.microsoft.com/en-us/dotnet/api/system.text.json?view=net-7.0

File: iwa-console/Program.cs [hunk 5]
```diff
@@ -33,14 +35,32 @@ static void Main(string[] args)
...
         {
-            SampleConfiguration config = SampleConfiguration.ReadFromJsonFile("appsettings.json");
-            var appConfig = config.PublicClientApplicationOptions;
-            var app = PublicClientApplicationBuilder.CreateWithApplicationOptions(appConfig)
-                                                    .Build();
-            var httpClient = new HttpClient();
-
-            MyInformation myInformation = new MyInformation(app, httpClient, config.MicrosoftGraphBaseEndpoint);
-            await myInformation.DisplayMeAndMyManagerAsync();
+            // Using appsettings.json as our configuration settings and utilizing IOptions pattern - https://learn.microsoft.com/dotnet/core/extensions/options
```
Comment (bgavrilMS) [inline iwa-console/Program.cs:38]: Please move all this logic into Main and use

`public static Task Main()`

File: iwa-console/MSAL/MSALClientHelper.cs [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+        /// </value>
+        public AuthenticationResult AuthResult { get; private set; }
+
+        /// <summary>
+        /// Gets a value indicating whether this instance of PublicClientApp was initialized with a broker .
+        /// </summary>
+        /// <value>
+        ///   <c>true</c> if this instance is broker initialized; otherwise, <c>false</c>.
+        /// </value>
+        public bool IsBrokerInitialized { get; private set; }
```
Comment (bgavrilMS) [inline iwa-console/MSAL/MSALClientHelper.cs:38]: I don't think we should expose this to end-users

File: iwa-console/MSAL/MSALClientHelper.cs [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+
+            await AttachTokenCache();
+            return await FetchSignedInUserFromCache().ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Initializes the public client application of MSAL.NET with the required information to correctly authenticate the user.
+        /// </summary>
+        /// <returns>An IAccount of an already signed-in user (if available)</returns>
+        public async Task<IAccount> InitializePublicClientAppForWAMBrokerAsync(IntPtr? handle)
```
Comment (bgavrilMS) [inline iwa-console/MSAL/MSALClientHelper.cs:90]: Can you just extract the handle value here instead of injecting it from somewhere else?

File: iwa-console/MSAL/MSALClientHelper.cs [hunk 8]
```diff
@@ -0,0 +1,270 @@
...
+            // Initialize the MSAL library by building a public client application for authenticating using WAM
+            this.PublicClientApplication = this.PublicClientApplicationBuilder
+                    .WithBrokerPreview(true)
+                    .WithParentActivityOrWindow(() => { return handle.Value; })                                     // Specify Window handle - (required for WAM).
+                    .Build();
+
+            this.IsBrokerInitialized = true;
+
+            await AttachTokenCache();
+            return await FetchSignedInUserFromCache().ConfigureAwait(false);
```
Comment (bgavrilMS) [inline iwa-console/MSAL/MSALClientHelper.cs:101]: Initialize should not do more than its name. It should not fetch users.

File: iwa-console/MSAL/MSALClientHelper.cs [hunk 9]
```diff
@@ -0,0 +1,270 @@
...
+        /// <returns>An IAccount of an already signed-in user (if available)</returns>
+        public async Task<IAccount> InitializePublicClientAppForWAMBrokerAsync(IntPtr? handle)
+        {
+            // Initialize the MSAL library by building a public client application for authenticating using WAM
+            this.PublicClientApplication = this.PublicClientApplicationBuilder
+                    .WithBrokerPreview(true)
+                    .WithParentActivityOrWindow(() => { return handle.Value; })                                     // Specify Window handle - (required for WAM).
+                    .Build();
+
+            this.IsBrokerInitialized = true;
```
Comment (bgavrilMS) [inline iwa-console/MSAL/MSALClientHelper.cs:98]: This is bad design, this flag can be set depending on the `Intialize` method used. Please remove flag and `InitializePublicClientAppAsync`

File: iwa-console/MSAL/MSALClientHelper.cs [hunk 10]
```diff
@@ -0,0 +1,270 @@
...
+        /// <summary>
+        /// Initializes the public client application of MSAL.NET with the required information to correctly authenticate the user.
+        /// </summary>
+        /// <returns>An IAccount of an already signed-in user (if available)</returns>
+        public async Task<IAccount> InitializePublicClientAppForWAMBrokerAsync(IntPtr? handle)
+        {
+            // Initialize the MSAL library by building a public client application for authenticating using WAM
+            this.PublicClientApplication = this.PublicClientApplicationBuilder
+                    .WithBrokerPreview(true)
+                    .WithParentActivityOrWindow(() => { return handle.Value; })                                     // Specify Window handle - (required for WAM).
```
Comment (bgavrilMS) [inline iwa-console/MSAL/MSALClientHelper.cs:95]: Set this only on Windows

Comment (bgavrilMS) [inline iwa-console/MSAL/MSALClientHelper.cs:95]: No need to use Func or .Value here. Just pass in the IntPtr i.e. `WithParentActivityOrWindow(handle)`

File: iwa-console/MSAL/MSALClientHelper.cs [hunk 11]
```diff
@@ -0,0 +1,270 @@
...
+            {
+                // 1. Try to sign-in the previously signed-in account
+                if (existingUser != null)
+                {
+                    this.AuthResult = await this.PublicClientApplication.AcquireTokenSilent(scopes, existingUser)
+                        .ExecuteAsync().ConfigureAwait(false);
+                }
+                else
+                {
+                    if (this.IsBrokerInitialized)
```
Comment (bgavrilMS) [inline iwa-console/MSAL/MSALClientHelper.cs:138]: Broker only please.

File: iwa-console/MSAL/MSALClientHelper.cs [hunk 12]
```diff
@@ -0,0 +1,270 @@
...
+                        Console.WriteLine("No accounts found in the cache. Trying Window's default account.");
+
+                        this.AuthResult = await this.PublicClientApplication
+                            .AcquireTokenSilent(scopes, Microsoft.Identity.Client.PublicClientApplication.OperatingSystemAccount)
+                            .ExecuteAsync()
+                            .ConfigureAwait(false);
+                    }
+                    else
+                    {
+                        this.AuthResult = await SignInUserInteractivelyAsync(scopes);
```
Comment (bgavrilMS) [inline iwa-console/MSAL/MSALClientHelper.cs:149]: Won't be needed in broker only, since `AcquireTokenSilent` will throw `MsalUiRequiredException `

**Comment by v-michaelmi** [comment]
@bgavrilMS 

Addressed the feedback you have mentioned. I updated the sample to use **NET7** and seems to work fine.

Only snag is the `WithParentActivityOrWindow` method only has signatures that accept `Func` parameters. Nothing seems to take a standalone `IntPtr`

**Comment by v-michaelmi** [comment]
Also added a trivial sign-out flow

**Comment by v-michaelmi** [comment]
@bgavrilMS running into a snag running this on Linux. If I use the `WithBroker` option it fails, and I don't believe we have support for Linux do we?

When I disable it, it works fine along with caching.

File: iwa-console/iwa-console.csproj (hunk 3 above)
Comment (v-michaelmi) [reply iwa-console/iwa-console.csproj]: It's not. That was an accidental addition.

File: iwa-console/MSAL/MSALClientHelper.cs (hunk 6 above)
Comment (v-michaelmi) [reply iwa-console/MSAL/MSALClientHelper.cs]: That can be refactored but are we supporting Brokers for `MacOS` and `Linux`?

File: iwa-console/MSAL/MSALClientHelper.cs (hunk 9 above)
Comment (v-michaelmi) [reply iwa-console/MSAL/MSALClientHelper.cs]: Based on what I see there is no signature that only takes an `IntPtr`. Needs to be wrapped in a function.

File: iwa-console/MSAL/MSALClientHelper.cs (hunk 10 above)
Comment (v-michaelmi) [reply iwa-console/MSAL/MSALClientHelper.cs]: Based on what I see there is no signature that only takes an IntPtr. Needs to be wrapped in a function.

Comment (bgavrilMS) [reply iwa-console/MSAL/MSALClientHelper.cs]: Ah ok yeah, you are right. That version is on AcquireTokenInteractive API. This is fine.

File: iwa-console/MSAL/MSALClientHelper.cs (hunk 6 above)
Comment (bgavrilMS) [reply iwa-console/MSAL/MSALClientHelper.cs]: No, but MSAL will fallback to a browser on Mac / Linux automatically. App developer doesn't need to worry about this.

**Comment by bgavrilMS** [comment]
> @bgavrilMS running into a snag running this on Linux. If I use the `WithBroker` option it fails, and I don't believe we have support for Linux do we?
> 
> When I disable it, it works fine along with caching.

Sorry, my bad, that API with IntPtr is on AcquireTokenInteractive. This API with Func<IntPtr> is on PCA object. This is fine./

**Comment by v-michaelmi** [comment]
@bgavrilMS That should do it. Works fine when I test it on Linux.

I have this setup so that it assumes the user wants to use **WAM** when on a windows device but let me know and I can add a flag for the user to choose if they want to use a browser instead.

If this LGTY I can make a repository or just rework what we have here. Whichever's easiest.

**Comment by v-michaelmi** [comment]
Obsolete.

Closing.
